### PR TITLE
Inventory selector for elevators

### LINKED ISSUE #2675: Book GUIs broken on latest build of CraftBukkit/Spiot/Paper
## :round_pushpin: Description (REQUIRED)

Elevator plates do not work and throw an error in the console "Could not pass event PlayerInteractEvent to Slimefun vDEV - 771 (git f2dd3a01)". This started after I updated Paper. It works fine in Paper Build 328 (and probably newer than that), but I have tried several of the "newer builds" and they do not work.

## :bookmark_tabs: Steps to reproduce the Issue (REQUIRED)

1. Update to a newer version of Paper
2. Step on a Slimefun Elevator plate

## :bulb: Expected behavior (REQUIRED)

Expect to be given option to go to another elevator level.

## :scroll: Server Log

https://pastebin.com/2b8EWn8Q

## :open_file_folder: /error-reports/ Folder

No error reports in Slimefun folder

## :compass: Environment (REQUIRED)

 - Server Software: Paper 1.16.4 Build 366
 - Minecraft Version: 1.16.4
 - Slimefun Version: 771
 - CS-CoreLib Version: 91
 - > sf versions
[07:52:17 INFO]: This Server uses the following setup of Slimefun:
[07:52:17 INFO]: Paper v1_16_R3
[07:52:17 INFO]: CS-CoreLib vDEV - 91 (git e42a05d1)
[07:52:17 INFO]: Slimefun vDEV - 771 (git f2dd3a01)
[07:52:17 INFO]: Metrics build: #6
[07:52:17 INFO]:
[07:52:17 INFO]: Installed Addons: (0)

### PR DESCRIPTION
**Pull request by WalshyDev**
## Description
Remove the book guide and the book interface from elevators.

*This is not ready for review, it is a draft. @TheBusyBiscuit and I are currently testing which approach is the best.*

Showcase video:
https://youtu.be/zxkeq8ePXys

## Changes
* Delete all book code
* Move the elevator code to a GUI
* Remove all traces of book guide

## Related Issues
Resolves #2675 

## Checklist

- [x] I have fully tested the proposed changes and promise that they will not break everything into chaos.
- [ ] I have also tested the proposed changes in combination with various popular addons and can confirm my changes do not break them.
- [x] I followed the existing code standards and didn't mess up the formatting.
- [ ] I did my best to add documentation to any public classes or methods I added.
- [x] I have added `Nonnull` and `Nullable` annotations to my methods to indicate their behaviour for null values
- [ ] I added sufficient Unit Tests to cover my code.

### CODE REVIEW

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/gps/ElevatorPlate.java [hunk 1]
```diff
@@ -119,45 +116,56 @@ public void openInterface(Player p, Block b) {
...
-
-        for (int i = 0; i < floors.size(); i++) {
-            if (i % 10 == 0) {
-                if (page != null) {
-                    book.addPage(page);
-                }
-
-                page = new ChatComponent(ChatColors.color(SlimefunPlugin.getLocalization().getMessage(p, "machines.ELEVATOR.pick-a-floor")) + "\n");
-            }
+    private void openFloorSelector(Block b, List<Block> floors, Player p, int page) {
```
Comment (J3fftw1) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/gps/ElevatorPlate.java:124]: annotations?

**Review by J3fftw1 — CHANGES_REQUESTED** [review]

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/gps/ElevatorPlate.java (hunk 1 above)
Comment (WalshyDev) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/gps/ElevatorPlate.java:124]: It has a `@ParametersAreNonnullByDefault` above

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/guide/SlimefunGuide.java [hunk 2]
```diff
@@ -104,14 +100,11 @@ public static void displayItem(PlayerProfile profile, SlimefunItem item, boolean
     }
 
     public static boolean isGuideItem(ItemStack item) {
-        return SlimefunUtils.isItemSimilar(item, getItem(SlimefunGuideLayout.CHEST), true) || SlimefunUtils.isItemSimilar(item, getItem(SlimefunGuideLayout.BOOK), true) || SlimefunUtils.isItemSimilar(item, getItem(SlimefunGuideLayout.CHEAT_SHEET), true);
+        return SlimefunUtils.isItemSimilar(item, getItem(SlimefunGuideLayout.CHEST), true)
+            || SlimefunUtils.isItemSimilar(item, getItem(SlimefunGuideLayout.CHEAT_SHEET), true);
     }
 
     public static SlimefunGuideLayout getDefaultLayout() {
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/guide/SlimefunGuide.java:107]: annotation :eyes: (and docs)

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/guide/SlimefunGuide.java [hunk 3]
```diff
@@ -104,14 +100,11 @@ public static void displayItem(PlayerProfile profile, SlimefunItem item, boolean
     }
 
     public static boolean isGuideItem(ItemStack item) {
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/guide/SlimefunGuide.java:102]: annotation :eyes:

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/guide/options/GuideLayoutOption.java [hunk 4]
```diff
@@ -1,115 +0,0 @@
...
-
-    @Override
-    public SlimefunAddon getAddon() {
-        return SlimefunPlugin.instance();
-    }
-
-    @Override
-    public NamespacedKey getKey() {
-        return new NamespacedKey(SlimefunPlugin.instance(), "guide_layout");
-    }
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/guide/options/GuideLayoutOption.java:31]: Removing this is a terrible idea... This way you effectively also removed the Cheat Sheet...

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/gps/ElevatorPlate.java [hunk 5]
```diff
@@ -119,45 +116,56 @@ public void openInterface(Player p, Block b) {
...
+                    Material.END_STONE,
+                    ChatColor.GRAY + "> " + (floors.size() - i) + ". " + ChatColor.BLACK + floor,
+                    SlimefunPlugin.getLocalization().getMessage(p, "machines.ELEVATOR.current-floor")
+                ));
             } else {
-                line = new ChatComponent("\n" + ChatColor.GRAY + (floors.size() - i) + ". " + ChatColor.BLACK + floor);
-                line.setHoverEvent(new HoverEvent(ChatColors.color(SlimefunPlugin.getLocalization().getMessage(p, "machines.ELEVATOR.click-to-teleport")), "", ChatColor.WHITE + floor, ""));
-                line.setClickEvent(new ClickEvent(new NamespacedKey(SlimefunPlugin.instance(), DATA_KEY + i), player -> teleport(player, floor, block)));
+                menu.addItem(i, new CustomItem(
+                    Material.END_STONE,
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/gps/ElevatorPlate.java:142]: I'd perhaps use a different material, maybe even numbered heads?

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/gps/ElevatorPlate.java [hunk 6]
```diff
@@ -119,45 +116,56 @@ public void openInterface(Player p, Block b) {
...
-
-                page = new ChatComponent(ChatColors.color(SlimefunPlugin.getLocalization().getMessage(p, "machines.ELEVATOR.pick-a-floor")) + "\n");
-            }
+    private void openFloorSelector(Block b, List<Block> floors, Player p, int page) {
+        ChestMenu menu = new ChestMenu("Floor Selector");
+        menu.setEmptySlotsClickable(false);
+        int pages = 1 + (floors.size() / 27);
 
+        for (int i = 0; i < Math.min(27, floors.size()); i++) {
+            // SlimefunPlugin.getLocalization().getMessage(p, "machines.ELEVATOR.pick-a-floor"))
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/gps/ElevatorPlate.java:130]: Why is this commented out?

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/gps/ElevatorPlate.java [hunk 7]
```diff
@@ -119,45 +116,56 @@ public void openInterface(Player p, Block b) {
...
-        for (int i = 0; i < floors.size(); i++) {
-            if (i % 10 == 0) {
-                if (page != null) {
-                    book.addPage(page);
-                }
-
-                page = new ChatComponent(ChatColors.color(SlimefunPlugin.getLocalization().getMessage(p, "machines.ELEVATOR.pick-a-floor")) + "\n");
-            }
+    private void openFloorSelector(Block b, List<Block> floors, Player p, int page) {
+        ChestMenu menu = new ChestMenu("Floor Selector");
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/gps/ElevatorPlate.java:125]: Why is this not localized? :P

**Review by TheBusyBiscuit — CHANGES_REQUESTED** [review]

**Pull request merged**